### DISCUSSION TITLE
Meaning of `Error: Unable to find any supported Python versions.` in Serverless Functions build.

### DISCUSSION THREAD

**Discussion by zhzhang** [vercel/vercel#11128] [Help] [answered]
### Question

I'm trying to deploy a python serverless function next to a NextJS app.
I have my script within `api/solver.py` and have `Pipfile` and `Pipfile.lock` in the project root, with python version 3.9.
The dependencies are `ortools` and `supabase`.

When I push to my branch, the preview build fails at `Error: Unable to find any supported Python versions.`
However, 3.9 is one of the supported versions.
Have I set up my project structure incorrectly? Do I need to enable python for serverless functions? Are the dependencies causing the issue?
The text of the error message is very opaque and nothing in the docs or previous discussions that I can find suggests a way to begin debugging this.

**Comment by dmarcucci** [comment]
Which version of Node are you using for your project? You can find it in Settings > General > Node.js Version.

I just switched a project from 18.x to 20.x as I'm trying to get ahead of the early 2025 deprecation date for 18.x.

In 18.x, I have no issue deploying my Edge functions alongside my Python endpoints. In 20.x, however, I get the same error you put above and adding `Pipfile` does not solve the issue.

**Reply by loderunner** [reply]
Same here. I had the same error, and switching back to Node 18.x fixed it.

**Reply by dmarcucci** [reply]
I'm actually trying to migrate these Python endpoints to JS (preferably Edge, but I'm not sure it's possible as it would require `fs` which I'm pretty sure Edge does not support), so that is a potential option for me to eventually switch to 20.x. However, it would be nice if there was some solution in the interim. And of course, other people may not be in position to switch their Python endpoints over.

**Reply by lappemic** [reply]
Can confirm that node<=18 worked for me as well.

**Reply by LilySu** [reply]
same node<=18 worked

**Reply by andrewcastle10** [reply]
Also confirming that switching to 18.x worked - thanks for the help y'all

**Reply by mchrisman** [reply]
I'm not using Node -- just a bare Python file -- so that workaround doesn't help.  AazimAnish's "config" section in vercel.json didn't fix it.

**Reply by dmarcucci** [reply]
@mchrisman

From what I've read elsewhere, it doesn't matter if you are using just Python - you need to set the Node version to 18.x regardless. This can either be done in `package.json` (per the workaround @amyegan linked to below) or in the settings for the project in the Vercel UI ("Settings" tab then scroll down to "Node.js Version" section and select "18.x" then do a new deploy). In your case, the latter probably makes most sense.

I agree that @AazimAnish comment is irrelevant.

**Reply by ndehouche** [reply]
Thank you for this. I had the same problem as OP and it fixed it. Still, this fix makes no sense...

**Comment by amyegan** [comment, marked as answer]
Thanks for reporting this. The engineering team is working on a bugfix now. But you can find a workaround posted in https://github.com/vercel/community/discussions/6287#discussioncomment-8932365

**Comment by MhdRaihan384** [comment]
{
    "builds": [{
        "src": "youtube_downloader/wsgi.py",
        "use": "@vercel/python",
        "config": { "maxLambdaSize": "15mb", "runtime": "python3.8" }
    }],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "youtube_downloader/wsgi.py"
        }
    ]
  }




when iam trying to deploy the django project in vercel it giveing me the erorr 

WARN! Due to `builds` existing in your configuration file, the Build and Development Settings defined in your Project Settings will not apply. Learn More: https://vercel.link/unused-build-settings
23:07:27.491 | Error: Unable to find any supported Python versions.
23:07:27.492 | Learn More: http://vercel.link/python-version

**Reply by Joshua-AS** [reply]
Have you found a solution, I'm having the same error

**Reply by dmarcucci** [reply]
This issue that @MhdRaihan384 describes is not related to the one mentioned above.

That being said, I think you need to set Python version to either 3.9 or 3.6 as those are the only 2 that Vercel supports at the moment (see [here](https://vercel.com/docs/functions/runtimes/python#python-version)).

3.6 is deprecated, so I would recommend 3.9.
